common/chan_utils_test.go

```diff
@@ -0,0 +1,190 @@
+/*
+ * Copyright 2024 Function Stream Org.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package common
+
+import (
+	"context"
+	"reflect"
+	"testing"
+	"time"
+)
+
+func TestSendToChannel(t *testing.T) {
+
+	// send data
+	t.Run("send_buffered_chan_success", func(t *testing.T) {
+		// buffered chan
+
+		c := make(chan string, 1)
+		ctx := context.Background()
+		if !SendToChannel(ctx, c, "data") {
+			t.Fatal("SendToChannel should return true when sending succeeds")
+		}
+		value := <-c
+		// Verify if the received data is correct
+		if value != "data" {
+			t.Errorf("expected to receive \"data\" from channel, but received %s", value)
+		}
+
+	})
+
+	t.Run("send_unbuffered_chan_success", func(t *testing.T) {
+		// unbuffered chan
+
+		c := make(chan string)
+		ctx := context.Background()
+
+		go func() {
+			SendToChannel(ctx, c, "data")
+		}()
+
+		value := <-c
+		// Verify if the received data is correct
+		if value != "data" {
+			t.Errorf("expected to receive \"data\" from channel, but received %s", value)
+		}
+
+	})
+
+	// context timeout
+	t.Run("context_not_timeout", func(t *testing.T) {
+		// test with a timeout setting, but test was successfully sent without timeout
+
+		c := make(chan string, 1)
+		ctx, cancel := context.WithTimeout(context.Background(), 100*time.Millisecond)
+		defer cancel()
+		if SendToChannel(ctx, c, "hello") {
+			t.Log("Data sent successfully")
+		} else {
+			t.Fatal("Failed to send data due to context timeout")
+		}
+		select {
+		case <-c:
+			t.Log("Successfully received data")
+		case <-ctx.Done():
+			t.Fatal("Fail to receive data due to the context timeout")
+		}
+
+	})
+
+	t.Run("context_timeout", func(t *testing.T) {
+		// time.Sleep setting timeout, simulating context timeout
+
+		c := make(chan string, 1)
+		ctx, cancel := context.WithTimeout(context.Background(), 100*time.Millisecond)
+		defer cancel()
+		if SendToChannel(ctx, c, "hello") {
+			t.Log("Data sent successfully")
+		} else {
+			t.Fatal("Failed to send data due to context timeout")
+		}
```

Comment:
We should check that when the context is timeout, the SendToChannel should return false.